dataframe-geo/src/main/kotlin/org/jetbrains/kotlinx/dataframe/geo/GeoDataFrame.kt

```diff
@@ -0,0 +1,66 @@
+package org.jetbrains.kotlinx.dataframe.geo
+
+import org.geotools.api.referencing.crs.CoordinateReferenceSystem
+import org.geotools.geometry.jts.JTS
+import org.geotools.referencing.CRS
+import org.jetbrains.kotlinx.dataframe.DataFrame
+import org.jetbrains.kotlinx.dataframe.api.update
+import org.jetbrains.kotlinx.dataframe.api.with
+
+/**
+ * A data structure representing a geographical DataFrame, combining spatial data with
+ * an optional Coordinate Reference System (CRS).
+ *
+ * @param T The type parameter extending `WithGeometry`, indicating the presence of a geometry column.
+ * @property df The underlying `DataFrame` containing geometries.
+ * @property crs The coordinate reference system associated with the data, if any.
+ */
+class GeoDataFrame<T : WithGeometry>(val df: DataFrame<T>, val crs: CoordinateReferenceSystem?) {
+    /**
+     * Updates the `GeoDataFrame` using a specified transformation block on the underlying DataFrame.
+     *
+     * @param updateBlock The block defining the transformations to be applied to the DataFrame.
+     * @return A new `GeoDataFrame` instance with updated data and the same CRS.
+     */
+    fun update(updateBlock: DataFrame<T>.() -> DataFrame<T>): GeoDataFrame<T> {
+        return GeoDataFrame(df.updateBlock(), crs)
+    }
+
+    /**
+     * Transforms the geometries to a specified Coordinate Reference System (CRS).
+     *
+     * This function reprojects the geometry data from the current CRS to a target CRS.
+     * If no target CRS is specified and the `GeoDataFrame` has no CRS, WGS 84 is used by default.
+     *
+     * @param targetCrs The target CRS for transformation.
+     * @return A new `GeoDataFrame` with reprojected geometries and the specified CRS.
+     */
+    fun applyCrs(targetCrs: CoordinateReferenceSystem): GeoDataFrame<T> {
+        if (crs == null) {
+            return GeoDataFrame(df, targetCrs)
+        }
+        if (targetCrs == this.crs) return this
+        // Use WGS 84 by default TODO
+        val sourceCRS: CoordinateReferenceSystem = this.crs ?: DEFAULT_CRS
+        val transform = CRS.findMathTransform(sourceCRS, targetCrs, true)
+        return GeoDataFrame(
+            df.update { geometry }.with { JTS.transform(it, transform) },
+            targetCrs
+        )
+    }
+
+    override fun equals(other: Any?): Boolean {
```

Comment:
missing hashCode and toString functions

Reply:
<img width="476" alt="image" src="https://github.com/user-attachments/assets/18b9dd12-f02a-4e12-8c3e-f8b777a88597">

Reply:
Well, they are redundant if you write them like that yeah, haha. I think you've gotten a bit used to `data class`'es and what they do for us.

Given `public data class Person(val name: String, val age: Int)`, Kotlin generates:

```kt
public class Person(public val name: String, public val age: Int) {
    // irrelevant for equality
    public operator fun component1(): String = name
    public operator fun component2(): Int = age
    public fun copy(name: String = this.name, age: Int = this.age): Person = Person(name, age)

    // for rendering the class neatly
    public override fun toString(): String = "Person(name=$name, age=$age)"

    // for checking equality based on the properties
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Person

        if (age != other.age) return false
        if (name != other.name) return false

        return true
    }

    // if two objects have equals() == true, then their hashCodes must be equal too
    override fun hashCode(): Int {
        var result = age
        result = 31 * result + name.hashCode()
        return result
    }
}
```
If you don't use a data class, and you want equality of an object to be defined by its properties, you'll have to write all of this manually. Sets, maps etc. often only check `hashCode()`, instead of the slower `equals()` to verify whether an element is already present.

`toString()`, `hashCode()`, and `equals()` can be generated by the IDE with Alt+Insert.